halo2_proofs/src/plonk/circuit.rs

```diff
@@ -706,6 +792,51 @@ pub enum Expression<F> {
 }
 
 impl<F: Field> Expression<F> {
+    /// Make side effects
+    pub fn query_cells(&mut self, cells: &mut VirtualCells<'_, F>) {
+        match self {
+            Expression::Constant(_) => (),
+            Expression::Selector(selector) => {
+                cells.queried_selectors.push(*selector);
+            }
+            Expression::Fixed(query) => {
+                let col = Column {
+                    index: query.column_index,
+                    column_type: Fixed,
+                };
+                cells.queried_cells.push((col, query.rotation).into());
+                query.index = Some(cells.meta.query_fixed_index(col, query.rotation));
```

Comment:
Following Han's comment here https://github.com/privacy-scaling-explorations/halo2/pull/154/files#r1127741332
We could wrap this over `if query.index.is_none() { ... }`.
The same applies for `Advice` and `Instance`.